flink-core/src/main/java/org/apache/flink/api/java/typeutils/RowTypeInfo.java

```diff
@@ -54,6 +76,152 @@ public RowTypeInfo(TypeInformation<?>... types) {
+			// handle select all
+			int keyPosition = 0;
+			for (TypeInformation<?> fType : types) {
+				if (fType instanceof CompositeType) {
+					CompositeType<?> cType = (CompositeType<?>) fType;
+					cType.getFlatFields(ExpressionKeys.SELECT_ALL_CHAR, offset + keyPosition, result);
+					keyPosition += cType.getTotalFields() - 1;
+				} else {
+					result.add(new FlatFieldDescriptor(offset + keyPosition, fType));
+				}
+				keyPosition++;
+			}
+		} else {
+			field = matcher.group(1);
+
+			Matcher intFieldMatcher = PATTERN_INT_FIELD.matcher(field);
+			TypeInformation<?> fieldType = null;
+			if (intFieldMatcher.matches()) {
+				// field expression is an integer
+				int fieldIndex = Integer.valueOf(field);
+				if (fieldIndex > this.getArity()) {
+					throw new InvalidFieldReferenceException(
+						"Row field expression \"" + field + "\" out of bounds of " + this.toString() + ".");
+				}
+				for (int i = 0; i < fieldIndex; i++) {
+					offset += this.getTypeAt(i).getTotalFields();
+				}
+				fieldType = this.getTypeAt(fieldIndex);
+			} else {
+				for (int i = 0; i < this.fieldNames.length; i++) {
+					if (fieldNames[i].equals(field)) {
+						// found field
+						fieldType = this.getTypeAt(i);
+						break;
+					}
+					offset += this.getTypeAt(i).getTotalFields();
+				}
+				if (fieldType == null) {
+					throw new InvalidFieldReferenceException(
+						"Unable to find field \"" + field + "\" in type " + this.toString() + ".");
+				}
+			}
+
+			String tail = matcher.group(3);
+
+			if (tail == null) {
+				// expression hasn't nested field
+				if (fieldType instanceof CompositeType) {
+					((CompositeType) fieldType).getFlatFields("*", offset, result);
+				} else {
+					result.add(new FlatFieldDescriptor(offset, fieldType));
+				}
+			} else {
+				// expression has nested field
+				if (fieldType instanceof CompositeType) {
+					((CompositeType) fieldType).getFlatFields(tail, offset, result);
+				} else {
+					throw new InvalidFieldReferenceException(
+						"Nested field expression \"" + tail + "\" not possible on atomic type " + fieldType + ".");
+				}
+			}
+		}
+	}
+
+	@Override
+	public <X> TypeInformation<X> getTypeAt(String fieldExpression) {
+		Matcher matcher = PATTERN_NESTED_FIELDS.matcher(fieldExpression);
+		if (!matcher.matches()) {
+			if (fieldExpression.equals(ExpressionKeys.SELECT_ALL_CHAR) ||
+				fieldExpression.equals(ExpressionKeys.SELECT_ALL_CHAR_SCALA)) {
+				throw new InvalidFieldReferenceException("Wildcard expressions are not allowed here.");
+			} else {
+				throw new InvalidFieldReferenceException("Invalid format of Row field expression \""+fieldExpression+"\".");
+			}
+		}
+
+		String field = matcher.group(1);
+
+		Matcher intFieldMatcher = PATTERN_INT_FIELD.matcher(field);
+		TypeInformation<X> fieldType = null;
+		if (intFieldMatcher.matches()) {
+			// field expression is an integer
+			int fieldIndex = Integer.valueOf(field);
+			if (fieldIndex > this.getArity()) {
```

Comment:
`TupleTypeBase.getFieldAt()` does check for index bounds as well and throws an `IndexOutOfBoundsException`.
So we could simplify this a bit here.